### ISSUE TITLE
Fine Tuning on Squad - Where to Lower Case?

### ISSUE THREAD

**Issue by ctmckee** [google-research/text-to-text-transfer-transformer#144]
I'm still geeking out on how cool these T5 and BERT are for Q&A, but I'm running into an issue for training accuracy with T5.

I trained the model using your pre-defined squad Task (below is how I defined the Task for registry).  The training accuracy at 25000 steps was: INFO:tensorflow:eval/squadTask/accuracy at step 1025000: 47.124.  ?

1. Does t5.data.preprocessor.squad, lower case the input or do I need to do that prior to the squad preprocessor? 
2. Does upper and lower case effect whether or not the prediction is Counted as Correct?
3. Finally I tested my model by using BERT and T5 to ask an article (context) what the authors concluded (question)?  The BERT response contained a string of words that was present in the context.  T5 formulated a string of words that was located nowhere in the context (even single words that were not in the context), but was relevant to the subject?  Will the base model of T5 produce higher probability responses from context not supplied directly with the question?

```python
t5.data.TaskRegistry.add(
    "squadTask",
    # A TfdsTask takes in a TFDS name instead of a tf.data.Dataset function.
    t5.data.TfdsTask,
    tfds_name="squad:1.0.0",
    tfds_data_dir=DATA_DIR,
    # Supply a function which preprocesses text from the tf.data.Dataset.
    text_preprocessor=[t5.data.preprocessors.squad],
    # Use the same vocabulary that we used for pre-training.
    sentencepiece_model_path=t5.data.DEFAULT_SPM_PATH,
    # Lowercase targets before computing metrics.
    postprocess_fn=t5.data.postprocessors.lower_text, 
    # We'll use accuracy as our evaluation metric.
    metric_fns=[t5.evaluation.metrics.accuracy],
    # Not required, but helps for mixing and auto-caching.
    #num_input_examples=num_nq_examples
)
```

**Comment by ctmckee** [comment]
Also,  here is how I supplied two questions and contexts to T5, based upon your example

```python
predict_inputs_path = os.path.join(MODEL_DIR, "predict_inputs_%d.txt" % now)
predict_outputs_path = os.path.join(MODEL_DIR, "predict_outputs_%d.txt" % now)
with tf.io.gfile.GFile(predict_inputs_path, "w") as f:
  f.write("question: %s" % question_1.lower()+'context:%s\n'%context_1.lower())
  f.write("question: %s" % question_2.lower()+'context:%s\n'%context_2.lower())

# Ignore any logging so that we only see the model's answers to the questions.
with tf_verbosity_level('ERROR'):
  model.batch_size = 2
  model.predict(
      input_file=predict_inputs_path,
      output_file=predict_outputs_path,
      # Select the most probable output token at each step.
      temperature=0,
  )

**Comment by craffel** [comment]
Hi, it looks like you're creating a new task for squad. I recommend you just use the squad task that is already in T5, because it has all of the appropriate preprocessing and evaluation: 
https://github.com/google-research/text-to-text-transfer-transformer/blob/master/t5/data/tasks.py#L283

> The BERT response contained a string of words that was present in the context. T5 formulated a string of words that was located nowhere in the context (even single words that were not in the context), but was relevant to the subject?

BERT can only extract spans out of the input when answering questions. T5 is a generative model that can generate arbitrary text. They take a fundamentally different approach. This is expected behavior.

**Comment by ctmckee** [comment]
OK, thanks I'll review the link you sent.  I thought I did use the squad task, by setting text_preprocessor = [t5.data.preprocessors.squad],  in the t5.data.TaskRegistry.add() shown above.

**Comment by craffel** [comment]
That sets the text preprocessor, but the task includes other factors such as the postprocessing and evaluation metrics.

**Comment by ctmckee** [comment]
Thanks for the support.  Here's the output I got accuracy output I acheived....which seems too high?  

INFO:absl:EM = 88.11, F1 = 94.49
INFO:tensorflow:eval/cmMod_squad_v010_allanswers/em at step 1010000: 88.108
INFO:tensorflow:eval/cmMod_squad_v010_allanswers/f1 at step 1010000: 94.488

If others are interested in trying to recreate.  I modified your squad with tfds_data_dir=DATA_DIR, because I don't know how to define that kwarg when setting the task_name in model.finetune().  Setting that DATA_DIR to a gs bucket location is required in colab, so you don't get a local directory error.

```python
#task
t5.data.TaskRegistry.add(
    "cmMod_squad_v010_allanswers",
    t5.data.utils.TfdsTask,
    tfds_name="squad/plain_text:1.0.0",
    tfds_data_dir=DATA_DIR,
    text_preprocessor=t5.data.preprocessors.squad,
    postprocess_fn=t5.data.postprocessors.qa,
    metric_fns=[t5.evaluation.metrics.squad],
    sentencepiece_model_path=t5.data.utils.DEFAULT_SPM_PATH
)

MODEL_SIZE = "3B" #@param["small", "base", "large", "3B", "11B"]
# Public GCS path for T5 pre-trained model checkpoints
# Public GCS path for T5 pre-trained model checkpoints
BASE_PRETRAINED_DIR = "gs://..."
PRETRAINED_DIR = os.path.join(BASE_PRETRAINED_DIR, MODEL_SIZE)
MODEL_DIR = os.path.join(MODELS_DIR, MODEL_SIZE)

if ON_CLOUD and MODEL_SIZE == "3B":
  tf.logging.warn(
      "The `3B` model is too large to use with the 5GB GCS free tier. "
      "Make sure you have at least 25GB on GCS before continuing."
  )
elif ON_CLOUD and MODEL_SIZE == "11B":
  raise ValueError(
      "The `11B` parameter is too large to fine-tune on the `v2-8` TPU "
      "provided by Colab. Please comment out this Error if you're running "
      "on a larger TPU."
  )

# Set parallelism and batch size to fit on v2-8 TPU (if possible).
# Limit number of checkpoints to fit within 5GB (if possible).
model_parallelism, train_batch_size, keep_checkpoint_max = {
    "small": (1, 256, 16),
    "base": (2, 128, 8),
    "large": (8, 64, 4),
    "3B": (8, 16, 1),
    "11B": (8, 16, 1)}[MODEL_SIZE]

tf.io.gfile.makedirs(MODEL_DIR)
# The models from our paper are based on the Mesh Tensorflow Transformer.
model = t5.models.MtfModel(
    model_dir=MODEL_DIR,
    tpu=TPU_ADDRESS,
    tpu_topology=TPU_TOPOLOGY,
    model_parallelism=model_parallelism,
    batch_size=train_batch_size,
    sequence_length={"inputs": 384, "targets": 32},
    learning_rate_schedule=0.001,
    save_checkpoints_steps=5000,
    keep_checkpoint_max=keep_checkpoint_max if ON_CLOUD else None,
    iterations_per_loop=100,
)

FINETUNE_STEPS = 10000 #@param {type: "integer"}

model.finetune(
    mixture_or_task_name="cmMod_squad_v010_allanswers",
    pretrained_model_dir=PRETRAINED_DIR,
    finetune_steps=FINETUNE_STEPS
)
```

**Comment by craffel** [comment]
For 3B, that is the expected score (see Table 14 in https://arxiv.org/pdf/1910.10683.pdf)
